### ISSUE TITLE
Cpython build failure: _bootstrap_python segfault

### ISSUE THREAD

**Issue by fthain** [python/cpython#127545]
# Bug report

### Bug description:

When building dev-lang/python-3.13.0 on Gentoo/m68k, the following crash is observed. 

```
./_bootstrap_python ./Programs/_freeze_module.py abc ./Lib/abc.py Python/frozen_modules/abc.h
gc: collecting generation 0...
make: *** [Makefile:1697: Python/frozen_modules/abc.h] Segmentation fault
```

A patch is available to resolve this issue. A pull request will follow shortly.

### CPython versions tested on:

3.13

### Operating systems tested on:

Linux

### Linked PRs
* gh-127546
* gh-135016
* gh-135209

**Comment by fthain** [comment]
Debian has a bug report for this issue.
[https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1105110](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1105110)

**Comment by AA-Turner** [comment]
Note m86k is not a supported platform per PEP 11.

**Comment by ZeroIntensity** [comment]
Reading the Debian reports, it seems like the plan is for m86k to switch to 4-byte alignment anyway. Do we need to fix this?

**Comment by fthain** [comment]
> Reading the Debian reports, it seems like the plan is for m86k to switch to 4-byte alignment anyway. Do we need to fix this?

@ZeroIntensity, there is no alternative ABI for Linux/m68k. That's why Debian picked up this patch.

No kernel and toolchain patches have been posted that would constitute a new ABI and no effort has been made to design a new ABI to meet the needs of everyone in the Linux/m68k community for the foreseeable future.

Distro maintainers are primarily concerned with their own software and its users, which is understandable.

Regardless, I think the patch stands on its merits because, IMO, the patch improves the python source code by improving correctness, readability, portability and reusability.

**Comment by chewi** [comment]
> Reading the Debian reports, it seems like the plan is for m86k to switch to 4-byte alignment anyway. Do we need to fix this?

Opinions on this do vary. 🙂 However, as someone who would like to see 4-byte alignment, I don't want to force it on anybody and would still like this fixed. Thank you for your efforts, @fthain.

**Comment by ZeroIntensity** [comment]
I'm worried that this might cause more problems with portability than solutions. `Py_ALIGNED`, as proposed in the PR, is a thin wrapper around GCC's `__attribute__((aligned(...)))`. We could use something like `_Py_ALIGN_AS`, but then all of a sudden we're requiring compilers to support C11 `_Alignas`. Allowing the object structure to have a different alignment based on the compiler does not sound fun.

**Comment by AA-Turner** [comment]
We also don't test this, as evidenced by that all buildbots & CI don't fail. I would suggest closing, this is an unsupported platform. Perhaps this could be re-evaluated with a BuildBot, but even then the concerns Peter raises remain.

A

**Comment by thesamesam** [comment]
> We could use something like _Py_ALIGN_AS, but then all of a sudden we're requiring compilers to support C11 _Alignas

Python already requires C11 in its headers.

Also, the bits in https://github.com/capi-workgroup/decisions/issues/61 make clear that it's fine as-needed.

**Comment by ZeroIntensity** [comment]
> Python already requires C11 in its headers.

Yes, but people notoriously don't always follow it (looking at you, Windows).

I think we should definitely get sign-off from the C API WG before requiring `_Py_ALIGN_AS` to build the core (especially for something as fundamental as `PyObject`).

**Comment by zooba** [comment]
> Python already requires C11 in its headers

It shouldn't, the headers are compatible with C++. `_Py_ALIGN_AS` is designed to be okay anywhere it's needed, but we'll find out very quickly indeed if it isn't sufficient once it appears in `PyObject`.

@encukou Thoughts here?

**Comment by encukou** [comment]
> Thoughts here?

If we want to introduce `_Py_ALIGN_AS` in the regular builds, this looks like the perfect issue to do it.

It is exactly the use case for `_Py_ALIGN_AS`.

If there are any issues with `PyObject`, we'll know quickly. If it turns out that we do need to revert, we “only” lose compatibility with an unsupported platform -- we'd be back to the drawing board on *this* issue.

I'd hold the 3.13 backport though; and maybe even 3.14 just to be safe.
People building for `m68k` can backport the patch downstream; in GCC-only land there shouldn't be much to worry about.


> I think we should definitely get sign-off from the C API WG before requiring _Py_ALIGN_AS to build the core

The sign-off is done. Judging by the [vote](https://github.com/capi-workgroup/decisions/issues/61), the WG *prefers* `_Py_ALIGN_AS` being in both builds. It just wasn't *needed*, and thus tested, in the regular one yet.

> We also don't test this [...] this is an unsupported platform.

Read “unsupported” more as “not in our CI” and less as “we won't spend time on this”.

We wouldn't magically add support for `m68k` with the patch; it's still on Gentoo/m68k & other maintainers to build, test, and report issues/PRs that we might merge.
The important thing for us is that supported platforms don't break.

---

*Thank you for supporting Python on platforms we can't test here!*

**Comment by encukou** [comment]
Now merged in `main`.
Generally, we don't backport fixes for unsupported architectures -- that would implicitly promise some support, and we can't test on m64k. Please patch locally for Debian & Gentoo. Does that sound fair?
Python 3.14 & below shouldn't change much any more, but if things break, we can help.

For GCC (and presumably clang), I'd recommend going with @fthain's simpler patch (using `Py_ALIGNED(1 << _PyObject_ALIGNMENT_SHIFT)`).

**Comment by fthain** [comment]
> Now merged in `main`. Generally, we don't backport fixes for unsupported architectures -- that would implicitly promise some support, and we can't test on m64k. Please patch locally for Debian & Gentoo. Does that sound fair? Python 3.14 & below shouldn't change much any more, but if things break, we can help.
> 
> For GCC (and presumably clang), I'd recommend going with [@fthain](https://github.com/fthain)'s simpler patch (using `Py_ALIGNED(1 << _PyObject_ALIGNMENT_SHIFT)`).

Thanks for your help in fixing this, @encukou.
For Debian, Gentoo, buildroot et al. I have now backported this fix on the [gc_crash-3.13](https://github.com/fthain/cpython/tree/gc_crash-3.13) branch. The [gc_crash](https://github.com/fthain/cpython/tree/gc_crash) branch is suitable for python 3.14.

**Comment by thesamesam** [comment]
Thanks all. @fthain can you file a bug on the Gentoo side for backporting please?
